internal/auth/issuer/pam/pam_issuer.go

```diff
@@ -0,0 +1,951 @@
+	if s.config.ClientSecret != "" {
+		// Confidential client - require client_secret_post authentication
+		if req.ClientSecret == nil || *req.ClientSecret != s.config.ClientSecret {
+			return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_client")}, nil
+		}
+	}
+	// For public clients (empty secret), we accept the request without secret validation
+	// In production, this should be combined with PKCE (code_challenge/code_verifier) for security
+
+	// Validate and retrieve authorization code
+	codeData, exists := s.codeStore.GetCode(*req.Code)
+	if !exists {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_grant")}, nil
+	}
+
+	// Validate that the client ID matches the stored code
+	if codeData.ClientID != *req.ClientId {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_grant")}, nil
+	}
+
+	// Get user information from NSS
+	systemUser, err := s.pamAuthenticator.LookupUser(codeData.Username)
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_grant")}, nil
+	}
+
+	// Get user groups for roles
+	groups, err := s.pamAuthenticator.GetUserGroups(systemUser)
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+	}
+
+	// Map groups to roles and extract organizations
+	roles := s.mapGroupsToRoles(groups)
+	organizations := s.extractOrganizations(groups)
+
+	// Create identity for token generation
+	tokenGenerationRequest := authn.TokenGenerationRequest{
+		Username:      codeData.Username,
+		UID:           codeData.Username,
+		Organizations: organizations,
+		Roles:         roles,
+		Audience:      []string{codeData.ClientID}, // Set audience to client ID from authorization request
+		Issuer:        s.config.Issuer,             // Set issuer
+	}
+	// Generate access token with proper expiry (1 hour)
+	accessToken, err := s.jwtGenerator.GenerateTokenWithType(tokenGenerationRequest, time.Hour, "access_token")
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+	}
+
+	// Create token response
+	tokenResponse := &pamapi.TokenResponse{
+		AccessToken: lo.ToPtr(accessToken),
+		TokenType:   lo.ToPtr(pamapi.Bearer),
+		ExpiresIn:   lo.ToPtr(int(time.Hour.Seconds())),
+	}
+
+	// Only generate refresh token if offline_access was requested
+	if strings.Contains(codeData.Scope, "offline_access") {
+		refreshToken, err := s.jwtGenerator.GenerateTokenWithType(tokenGenerationRequest, 7*24*time.Hour, "refresh_token")
+		if err != nil {
+			return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+		}
+		tokenResponse.RefreshToken = lo.ToPtr(refreshToken)
+	}
+
+	return tokenResponse, nil
+}
+
+// Authorize handles the authorization endpoint for authorization code flow
+func (s *PAMOIDCProvider) Authorize(ctx context.Context, req *pamapi.AuthAuthorizeParams) (*AuthorizeResponse, error) {
+	s.log.Infof("Authorize: starting authorization request - ClientId=%s, RedirectUri=%s, ResponseType=%s, Scope=%s, State=%s",
```

Comment:
question: how chatty is this should it be debug?

Reply:
fixed